### ISSUE TITLE
Troubleshoot CUDA not being detected

### ISSUE THREAD

**Issue by TriceHelix** [m4rs-mt/ILGPU#912]
I added ILGPU to a WPF project in Visual Studio, however when creating a new Context the only device that appears is the CPU accelerator (when CUDA should be there too). Since there does not seem to be an API for "detecting" other devices, this is either a bug or an error on my end. I've already been having great success with ILGPU inside of Unity Engine, where everything works flawlessly. I have no idea how to troubleshoot this issue; I've already experimented with creating a context after a few seconds delay and updating my graphics drivers, yet it works in the Unity environment while refusing to do so in a WPF project. Is there anything I am completely overlooking?

**Comment by MoFtZ** [comment]
hi @TriceHelix. In order to use ILGPU, the WPF application needs to be compiled using the same bitness as the OS. So if you have a 64-bit OS, you need to compile as a 64-bit WPF application. There should be a `Prefer 32-bit` setting in the WPF project settings that needs to be turned off.

Also note that Nvidia removed support for 32-bit Cuda drivers.

**Comment by TriceHelix** [comment]
That totally makes sense, however I have double-checked my settings and this should all already be taken care of.
![config](https://user-images.githubusercontent.com/81759026/211198009-507bd9f1-4865-4a77-a302-25bfd069dc04.png)
This is where the option you mentioned used to be, I believe it was replaced by a dropdown where one can specify which exact configuration to prefer. I have `x64` selected.

The same configuration is selected for the project in the Configuration Manager (aswell as at the top of Visual Studio where you can specify Debug/Release etc). When I call `Environment.Is64BitProcess`, it returns `true`, yet nothing has changed about my issue. I am lost.

**Comment by TriceHelix** [comment]
UPDATE:
I have found the issue! Turns out things are complicated:
When running the application within Visual Studio, only DLLs and symbols required for the current executing code are loaded. WPF, Windows API, all that stuff for displaying the window is loaded immediately, but nothing else. Creating my first `ILGPU.Context` instance triggers the load of the `ILGPU.Runtime` DLL, meaning that regardless of how much delay there is between starting the application and creating a context (or accessing any ILGPU functionality, for that matter), the instruction is executed immediately upon ILGPU being loaded.

I assume that ILGPU actually takes some time to detect additional devices besides the CPU as soon as it is loaded, because **CUDA gets detected after a few seconds** if I create additional contexts after the first one. The reason this does not happen within Unity Engine is because they have a system in place which reloads the entire domain (consisting of every assembly and plugin) as soon as the application starts, unlike Visual Studio when debugging.

@MoFtZ Personally, I'd suggest putting some kind of logic in place to ensure that as soon as an `ILGPU.Context` is created, all devices available are detected. There is probably already an asynchronous operation ongoing which does the detection as soon as ILGPU is loaded, however there is no blocking done in the context creation which waits until the detection finishes. Then again, I don't know nearly enough about the black magic you use within ILGPU to make this amazing piece of software work.

TLDR for anyone else encountering this issue; create a context, block the thread for some time, dispose the context and create another one. This one should have all available devices.

**Comment by MoFtZ** [comment]
@TriceHelix you shouldn't need to do any "magic" like what you are describing.

What version of ILGPU are you using? And how are you creating the ILGPU context?

I have attached a sample WPF application using .NET 6. When the ILGPU context is created in the App constructor, or the MainWindow Load method, the Cuda accelerators are detected without issue.

[WpfApp1.zip](https://github.com/m4rs-mt/ILGPU/files/10377501/WpfApp1.zip)

**Comment by TriceHelix** [comment]
@MoFtZ A little delayed, but the flu hit me hard.

It turns out I made a silly mistake. I forgot to include a call to `Context.Builder.AllAccelerators()` in my context builder. Until recently I always used the CreateDefault() method you used in the attached project and then changing some settings afterwards (like a different optimization level). Now that I'm manually building the context I should make sure to think twice about any issues that arise.
I still have no idea why creating multiple Contexts unintentionally solved the issue (without actually explicitly enabling all Accelerators), but frankly that's something I don't wanna wrap my head around.

Apologies for my ignorance, and thank you for the help!
